src/conn/conn_dhandle.c

```diff
@@ -708,6 +708,15 @@ __wt_conn_dhandle_discard(WT_SESSION_IMPL *session)
 	__wt_session_close_cache(session);
 	F_SET(session, WT_SESSION_NO_DATA_HANDLES);
 
+	/*
+	 * The connection may have an open metadata cursor handle. We cannot
+	 * close it before now because it's potentially used when discarding
+	 * other open data handles. Close it before discarding the underlying
+	 * metadata handle.
+	 */
+	if (session->meta_cursor != NULL)
+		WT_TRET(session->meta_cursor->close(session->meta_cursor));
+
```

Comment:
I think that gets messy -- but I may just not understand the plan.

First, `__session_close` has a separate loop that closes open cursors before it calls `__wt_session_close_cache`, and so we'd attempt to close the metadata cursor twice, so I think there's an issue there.

Second, `__wt_conn_dhandle_discard` calls `__wt_session_close_cache` twice (before and after closing other open data handles), so wouldn't  that mean we'd close the metadata cursor, then close some other handle that would re-open the metadata cursor, and then close it again?